### PR TITLE
Validator key manifests

### PR DESCRIPTION
**Pull request by josh-ripple**
* Fix some things
* Add the `manifest.py` command-line tool for creating new config entries
* Support ephemeral validator keys in UniqueNodeList
* Add new ManifestCache
* Trade validator key manifests via peer gossip

### CODE REVIEW

**Comment by nbougalis** [comment]
Tag reviewers please.

**Comment by josh-ripple** [comment]
Reviewers: @scottschurr @ximinez

**Comment by josh-ripple** [comment]
I see a few issues remaining.  Is a revoking manifest supposed to include a new ephemeral signing key that will be ignored?  Because right now that's required.

**Comment by josh-ripple** [comment]
Is it really necessary to `verify()` before calling `maybe_insert()`?  Because `maybe_insert()` calls `unpackManifest()`, which calls `verify()`.  And why have a separate `would_accept()`?

**Comment by scottschurr** [comment]
A quick scan through suggests the bulk of the code here is in python.  Does it make sense to have @rec as a reviewer?

**Comment by nbougalis** [comment]
I definitely think that @rec should be tagged as a reviewer on this because of the Python bits.

**Comment by vinniefalco** [comment]
There's some commit bloat going on here

**Comment by ximinez** [comment]
Build failure in Windows:
```
cl /Fobuild\msvc.debug\src\ripple\unity\overlay.obj /c src\ripple\unity\overlay. cpp /TP /nologo /bigobj /EHa /fp:precise /Gd /Gm- /GR /Gy- /FS /MP /openmp- /Zc:forScope /Zi /errorReport:none /nologo /W3 /WX- /wd"4018" /wd"4244" /wd"4267" /wd"4800" /GS /MTd /Od /RTC1 /DOPENSSL_NO_SSL2 /DDEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER /DHAVE_USLEEP=1 /DDEBUG /D_DEBUG /D_WIN32_WINNT=0x6000 /D_SCL_SECURE_NO_WARNINGS /D_CRT_SECURE_NO_WARNINGS /DWIN32_CONSOLE /D_CRTDBG_MAP_ALLOC /IC:\boost_1_57_0 /IC:\OpenSSL-Win64\include /Isrc /Isrc\beast /Ibuild\proto /Isrc\soci\src /Isrc\protobuf\src /Isrc\protobuf\vsprojects
overlay.cpp
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xstring(1271) : error C2280: 'ripple::AnyPublicKey &ripple::AnyPublicKey::operator =(const ripple::AnyPublicKey &)' : attempting to reference a deleted function
        c:\Dev\rippled-merge\src\ripple/protocol/AnyPublicKey.h(105) : see declaration of 'ripple::AnyPublicKey::operator ='
        This diagnostic occurred in the compiler generated function 'ripple::Manifest &ripple::Manifest::operator =(const ripple::Manifest &)'
scons: *** [build\msvc.debug\src\ripple\unity\overlay.obj] Error 2
scons: building terminated because of errors.
```
I'll bet you're expecting a move assignment here, but, IIRC, msvc is terribly negligent about auto-creating move constructors and operators.

**Comment by josh-ripple** [comment]
@scottschurr No, `VER_NODE_PUBLIC` is also used in `UniqueNodeList.cpp`.

**Comment by josh-ripple** [comment]
Adding @rec as a reviewer for the Python bits.  (And feel free to comment on the rest as well, if you like.)

**Comment by josh-ripple** [comment]
@scottschurr Oops, `VER_NODE_PUBLIC` *should* be used in `Manifest.cpp`.  It just isn't.

**Comment by josh-ripple** [comment]
Passing a journal by const ref works as long as the object is in scope.  This one belongs to the overlay.

**Comment by josh-ripple** [comment]
I renamed `mIn` and mOut` to `inbox` and `outbox`.

**Comment by vinniefalco** [comment]
As long as we're renaming why not `in` and `out`

**Comment by josh-ripple** [comment]
Because saving three bytes/columns per use isn't worth the loss of descriptiveness, IMHO.

**Comment by vinniefalco** [comment]
`inbox` is no more descriptive than `in`

**Comment by ximinez** [comment]
Windows scons builds succeed and tests pass with https://github.com/josh-ripple/rippled/commit/14f90fbcc388abcbe1fa72f8f80881acea5e0aa4. 

The `.vcxproj` files are out of date.

File: bin/manifest.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+from ripple import ed25519
+from ripple import ecdsa
+from hashlib import sha256
+
+#-----------------------------------------------------------
+
+# 58 character alphabet used for Ripple
+human_alphabet = 'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'
+
+def base_encode(b, digits, check = 1):
```
Comment (rec) [inline bin/manifest.py]: No spaces around = in default parameters, according to PEP 8.

File: bin/manifest.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    """ Returns binary as a radix-encoded string """
+    lb = b.lstrip(b'\0')
+    bn = long(0)
+    for c in lb:
+        bn = bn * 256 + ord(c)
+    radix = len(digits)
+    s = ''
+    while (bn > 0):
```
Comment (rec) [inline bin/manifest.py]: no parentheses around the condition (PEP 8 again).

File: bin/manifest.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+#-----------------------------------------------------------
+
+# 58 character alphabet used for Ripple
+human_alphabet = 'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'
+
+def base_encode(b, digits, check = 1):
+    """ Returns binary as a radix-encoded string """
+    lb = b.lstrip(b'\0')
+    bn = long(0)
```
Comment (rec) [inline bin/manifest.py]: There's no need to cast to a long, Python will automatically upcast it to you if it needs to.

File: bin/manifest.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+import base64, os, random, sys
+from ripple import ed25519
+from ripple import ecdsa
+from hashlib import sha256
+
+#-----------------------------------------------------------
+
+# 58 character alphabet used for Ripple
+human_alphabet = 'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'
```
Comment (rec) [inline bin/manifest.py]: Constant names are uppercase:  https://www.python.org/dev/peps/pep-0008/#constants

File: bin/manifest.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    """ Returns binary as a radix-encoded string """
+    lb = b.lstrip(b'\0')
+    bn = long(0)
+    for c in lb:
+        bn = bn * 256 + ord(c)
+    radix = len(digits)
+    s = ''
+    while (bn > 0):
+        bn, mod = divmod(bn, radix)
+        s += digits[mod]
```
Comment (rec) [inline bin/manifest.py]: This is an anti-pattern - because it's quadratic in the number of parts.

Prefer

    sl = []
    while bn:
        sl.append(digits[mod])
     s = ''.join(sl)

File: bin/manifest.py (hunk 1 above)
Comment (rec) [inline bin/manifest.py]: default value should be `True`, not `1`

File: bin/manifest.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+from ripple import ecdsa
+from hashlib import sha256
+
+#-----------------------------------------------------------
+
+# 58 character alphabet used for Ripple
+human_alphabet = 'rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz'
+
+def base_encode(b, digits, check = 1):
+    """ Returns binary as a radix-encoded string """
```
Comment (rec) [inline bin/manifest.py]: No space after `"""`, period and no space at end.

File: bin/manifest.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    s = digits[0] * (len(b) - len(lb)) + s[::-1]
+    if (check):
+        assert base_decode(s, digits, 0) == b
+    return s
+
+def base_decode(s, digits, check = 1):
+    """ Returns the binary from decoding a radix-encoded string """
+    bn = 0
+    radix = len(digits)
+    ls = s.lstrip(str(digits[0]))
```
Comment (rec) [inline bin/manifest.py]: Why are you converting to a string?  Isn't it already a string?

File: bin/manifest.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    bn = long(0)
+    for c in lb:
+        bn = bn * 256 + ord(c)
+    radix = len(digits)
+    s = ''
+    while (bn > 0):
+        bn, mod = divmod(bn, radix)
+        s += digits[mod]
+    s = digits[0] * (len(b) - len(lb)) + s[::-1]
+    if (check):
```
Comment (rec) [inline bin/manifest.py]: no parens.

File: bin/manifest.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        assert base_decode(s, digits, 0) == b
+    return s
+
+def base_decode(s, digits, check = 1):
```
Comment (rec) [inline bin/manifest.py]: Same comments as the above!

File: bin/manifest.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    b = ''
+    while bn > 0:
+        bn, m = divmod(bn, 256)
+        b += chr(m)
+    b = b'\0' * (len(s) - len(ls)) + b[::-1]
+    if (check):
+        assert base_encode(b, digits, 0) == s
+    return b
+
+def base58encode(b):
```
Comment (rec) [inline bin/manifest.py]: Do you ever use the two functions above?  why not just make the default for `digits` to be `human_alphabet` in the above function?

File: bin/manifest.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python
+
+import base64, os, random, sys
+from ripple import ed25519
+from ripple import ecdsa
+from hashlib import sha256
+
```
Comment (rec) [inline bin/manifest.py]: Unit tests?

Getting base arithmetic right is prone to error, particularly on edge cases.

File: bin/manifest.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+VER_NODE_PUBLIC       = 28
+VER_NODE_PRIVATE      = 32
+VER_ACCOUNT_ID        = 0
+VER_ACCOUNT_PUBLIC    = 35
+VER_ACCOUNT_PRIVATE   = 34
+VER_FAMILY_GENERATOR  = 41
+VER_FAMILY_SEED       = 33
+
+def ra_check(b):
+    """ Returns a 4-byte checksum of a binary """
```
Comment (rec) [inline bin/manifest.py]: Same comment about comments as above - also applies below.

File: bin/manifest.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+
+#
+# RippleAddress library
+#
+# Human strings are base-58 with a
+# version prefix and a checksum suffix.
+#
+
+VER_NONE              = 1
```
Comment (rec) [inline bin/manifest.py]: What are these constants?  Why this order (neither alphabetic nor in order of the numbers...)?

File: bin/manifest.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+STI_UINT32          = 2
+STI_VL              = 7
+
+sfSequence          = field_code(STI_UINT32, 4)
+sfPublicKey         = field_code(STI_VL, 1)
+sfSigningPubKey     = field_code(STI_VL, 3)
+sfSignature         = field_code(STI_VL, 6)
+
+def to_bytes(n, length, endianess='big'):
```
Comment (rec) [inline bin/manifest.py]: endianness has two n's:  http://en.wikipedia.org/wiki/Endianness

File: bin/manifest.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+    h = '%x' % n
+    s = ('0'*(len(h) % 2) + h).zfill(length*2).decode('hex')
+    return s if endianess == 'big' else s[::-1]
+
+def lenvl(b):
```
Comment (rec) [inline bin/manifest.py]: This is pretty cryptic, perhaps a better name?

File: bin/manifest.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+    s = ''
+    n = len(b)
+    if (n < 192):
+        s += chr(n)
+        return s
+    raise Exception('too long')
+
+def strvl(b):
```
Comment (rec) [inline bin/manifest.py]: Is this worth it?

File: bin/manifest.py [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+    return lenvl(b) + b
+
+def str32(n):
```
Comment (rec) [inline bin/manifest.py]: Or this?

File: bin/manifest.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+def gen_ec():
+    sk = ecdsa.SigningKey.generate(curve=ecdsa.SECP256k1)
+    vk = sk.get_verifying_key()
+    sig = sk.sign("message")
+    assert vk.verify(sig, "message")
+    return sk, vk
+
+def gen_manifest(pk, vpk, seq):
+    s = ""
```
Comment (rec) [inline bin/manifest.py]: 1. prefer `'` over `"` (no shift, less "ink").
2. you're doing the append antipattern again!  :-)

File: bin/manifest.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+    s += sfPublicKey
+    s += strvl(pk)
+    s += sfSigningPubKey
+    s += strvl(vpk)
+    return s
+
+def sign_manifest(m, sk, pk):
+    s = "MAN\0"
+    s += m
+    sig = ed25519.signature(s, sk, pk)
```
Comment (rec) [inline bin/manifest.py]: Why not `sig =  ed25519.signature('MAN\0' + m, sk, pk)`?

File: bin/manifest.py [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+
+#-----------------------------------------------------------
+
+def wrap(s, cols = 60):
+    n = len(s)
+    if (n <= cols):
+        return s
+    l = (n + cols - 1) / cols
+    w = n / l
+    while(l > 0):
```
Comment (rec) [inline bin/manifest.py]: space after while, no parents (as always)

File: bin/manifest.py [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+    s = "MAN\0"
+    s += m
+    sig = ed25519.signature(s, sk, pk)
+    m += sfSignature + strvl(sig)
+    return m
+
+#-----------------------------------------------------------
+
+def wrap(s, cols = 60):
+    n = len(s)
```
Comment (rec) [inline bin/manifest.py]: `len()` is very cheap, this local variable isn't really worth it.

File: bin/manifest.py [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+        print ("Bad master-secret: " + skh)
+    pk = ed25519.publickey(sk)
+    apk = chr(0xed) + pk
+    m = gen_manifest(apk, avpk, seq)
+    m1 = sign_manifest(m, sk, pk)
+    print ('[validation_manifest]')
+    print wrap(base64.b64encode(m1))
+    exit()
+
+print("""\
```
Comment (rec) [inline bin/manifest.py]: Very common (good) pattern for long strings.

File: bin/manifest.py [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+if (len(sys.argv) == 5 and sys.argv[1] == 'sign'):
+    seq = int(sys.argv[2])
+    vpkh = sys.argv[3]
+    skh = sys.argv[4]
+    try:
+        v, avpk = ra_decode(vpkh)
+        if (v != VER_NODE_PUBLIC or
+            len(avpk) != 33 or
+            (ord(avpk[0]) != 2 and ord(avpk[0]) != 3)):
+            raise ValueError()
```
Comment (rec) [inline bin/manifest.py]: Why not actually raise the error, and not catch it?  it's a failure, yes?

File: bin/manifest.py [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+    except ValueError:
+        print ("Bad validator-public: " + vpkh)
+        exit()
+    try:
+        v, sk = ra_decode(skh)
+        if (v != VER_NODE_PRIVATE or
+            len(sk) != 32):
+            raise ValueError()
+    except ValueError:
+        print ("Bad master-secret: " + skh)
```
Comment (rec) [inline bin/manifest.py]: You can have multiple arguments to print, if it's a function:

    print('Bad master-secret:', skh)

avoids the extra string creation.

If your `print` is not a function, then it'll be:

    print 'Bad master-secret:', skh

File: bin/ripple/ecdsa/__init__.py [hunk 25]
```diff
@@ -0,0 +1,14 @@
+__all__ = ["curves", "der", "ecdsa", "ellipticcurve", "keys", "numbertheory",
+           "test_pyecdsa", "util", "six"]
+from .keys import SigningKey, VerifyingKey, BadSignatureError, BadDigestError
+from .curves import NIST192p, NIST224p, NIST256p, NIST384p, NIST521p, SECP256k1
+
+_hush_pyflakes = [SigningKey, VerifyingKey, BadSignatureError, BadDigestError,
+                  NIST192p, NIST224p, NIST256p, NIST384p, NIST521p, SECP256k1]
+del _hush_pyflakes
```
Comment (rec) [inline bin/ripple/ecdsa/__init__.py]: Huh?  What's going on here??

File: bin/manifest.py (hunk 15 above)
Comment (vinniefalco) [inline bin/manifest.py]: Its the lenth of a VL record in rippled's serialization format

File: bin/manifest.py (hunk 16 above)
Comment (vinniefalco) [inline bin/manifest.py]: What's being asked?

File: bin/manifest.py [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+def strvl(b):
+    return lenvl(b) + b
+
+def str32(n):
+    return to_bytes(n, 4)
+
+#-----------------------------------------------------------
+
+def gen_seed():
+    seed = os.urandom(16)
```
Comment (rec) [inline bin/manifest.py]: You really don't need this local variable!

File: bin/manifest.py (hunk 18 above)
Comment (vinniefalco) [inline bin/manifest.py]: Habits, brother! They die hard

File: bin/manifest.py [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+    return seed
+
+def gen_ed():
+    sk = os.urandom(32)
```
Comment (rec) [inline bin/manifest.py]: Nor these!

File: bin/manifest.py [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+    sk, pk = gen_ed()
+    apk = chr(0xed) + pk
+    pkh = ra_encode(VER_NODE_PUBLIC, apk)
+    skh = ra_encode(VER_NODE_PRIVATE, sk)
+    v0, apk0 = ra_decode(pkh)
+    assert v0 == VER_NODE_PUBLIC
+    assert apk0 == apk
+    print ("[validators]")
+    print (pkh)
+    print
```
Comment (rec) [inline bin/manifest.py]: Is your `print` a function or not?

Often, people put `from __future__ import print_function` at the very top of the file to enforce "print is a function" (from Python 3.0).

File: bin/manifest.py (hunk 23 above)
Comment (vinniefalco) [inline bin/manifest.py]: Just to make the message a bit more friendly

File: bin/ripple/ecdsa/__init__.py (hunk 25 above)
Comment (vinniefalco) [inline bin/ripple/ecdsa/__init__.py]: This is 3rd party code

File: bin/manifest.py [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+    l = (n + cols - 1) / cols
+    w = n / l
+    while(l > 0):
+        s = s[:w * l] + '\n' + s[w*l:]
+        l -= 1
+    return s
+
+#-----------------------------------------------------------
+
+if (len(sys.argv) == 2 and sys.argv[1] == 'create'):
```
Comment (rec) [inline bin/manifest.py]: 1. Consider putting this into a function, not at the top level.  You've just created a lot of global variables, which is bad practice.

2. This way of argument checking doesn't give clear results.

I can think of two neater ways, particularly if you expect new features later:

    args = sys.argv[1:]
    command = args.pop()
    if command == 'create':
        assert len(args) == 0, 'command takes no arguments'

or this fancier version, perhaps only for your own edification.

    def manifest():
        def create():
            "do stuff"

         def check(version):
             "do other stuff"

         try:
             locals()[sys.argv[1]](*sys.argv[2:])
         except KeyError:
             raise ValueError('Didn\'t understand command ' + sys.argv[1])
         except TypeError:
             raise ValueError('Wrong number of arguments for command ' + sys.argv[1])

File: bin/manifest.py [hunk 30]
```diff
@@ -0,0 +1,244 @@
...
+    sk, pk = gen_ed()
+    apk = chr(0xed) + pk
```
Comment (rec) [inline bin/manifest.py]: I've been guessing what these variables mean, but here I have reached a block... :-)

File: bin/manifest.py [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+    pkh = ra_encode(VER_NODE_PUBLIC, apk)
+    skh = ra_encode(VER_NODE_PRIVATE, sk)
```
Comment (rec) [inline bin/manifest.py]: Move the declaration of this to right before it actually gets used.  You might not even need the local variable.  Or if you do, give it a useful name please!

File: bin/manifest.py [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+    v0, apk0 = ra_decode(pkh)
+    assert v0 == VER_NODE_PUBLIC
+    assert apk0 == apk
+    print ("[validators]")
+    print (pkh)
+    print
+    print ("[master_secret]")
+    print (skh)
+    exit()
```
Comment (rec) [inline bin/manifest.py]: Ouch!  Generally, calling exit is considered a bit abrupt.  :-D

File: bin/ripple/ecdsa/__init__.py [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+
+# This code comes from http://github.com/warner/python-ecdsa
+
+from ._version import get_versions
+__version__ = get_versions()['version']
+del get_versions
```
Comment (rec) [inline bin/ripple/ecdsa/__init__.py]: That's a bit obsessive - though I've realize that this is some external code so I shouldn't critique it too much.

The pythonic way, if you're worried about exposing global variables (which is not terrible, though most people don't bother, on the theory that people simply shouldn't use stuff they aren't told to use): `from ._version import get_versions as _get_versions`

**Comment by rec** [comment]
We're importing someone else's code here - but it's in the same directory as our code.  

This isn't a good idea for a lot of reasons, if only because I'm not entirely sure which is existing code that can be more or less rubber stamped and which is new code that I need to fully review.

Using `sys.path`, you can easily make have the third-party code in a separate subdirectory and still have appear where it needs to in the Python dependencies.  But it looks as if they use entirely relative paths, so you can probably just drop it into a subdirectory!


I'm also starting to wonder whether we shouldn't be pulling out this Python as a separate repository?

**Comment by rec** [comment]
(Oh, and I concentrated on the Python because the C++ seemed to be well in hand.  let me know if you want comments on the C++...)

File: bin/manifest.py (hunk 31 above)
Comment (vinniefalco) [inline bin/manifest.py]: Abbreviations mean:
secret key, public key, any public key, public key human, secret key human

File: bin/manifest.py (hunk 30 above)
Comment (vinniefalco) [inline bin/manifest.py]: `AnyPublicKey`

File: bin/manifest.py (hunk 32 above)
Comment (rec) [inline bin/manifest.py]: if you made this a function, which you might want to, a `return` would be fine...

File: bin/manifest.py (hunk 23 above)
Comment (rec) [inline bin/manifest.py]: not Pythonic!  

always throw with a nice error, it does the right thing nearly all the time.  the people who are using this are going to be engineers, they want to see the stack trace you just trashed... :-)

File: bin/ripple/ecdsa/__init__.py (hunk 25 above)
Comment (rec) [inline bin/ripple/ecdsa/__init__.py]: Sure, I got that.  but that's still really peculiar.  did we get this right?

Comment (rec) [inline bin/ripple/ecdsa/__init__.py]: What I mean is that it's _provable_ that removing lines 6 through 8 would have no discernable effect on the program.  Items don't know that they're being added to a list;  making a list has no side effects;  the list is immediately deleted.

I just want to make sure that this isn't some editing or copying error somewhere.

**Comment by rec** [comment]
The third party code is in the wrong subdirectory!

Putting it in the `ripple` namespace implies its our code, when it isn't. And if we used some other third party library that also used ecdsa, it wouldn't work.

Unfortunately, our Python root directory is in `bin/`.  I didn't like that at the time.

Three choices

* Put this code in `bin/ecdsa`
* Put this code in `bin/third_party/ecdsa` and then do some magic with `sys.path`.
* Move our Python to `bin/python/ripple` and have this as `bin/python/ecdsa`.

File: bin/ripple/ecdsa/__init__.py (hunk 25 above)
Comment (vinniefalco) [inline bin/ripple/ecdsa/__init__.py]: hmm....I don't have a really good answer for you ^_^ perhaps we might ask the authors.

File: bin/manifest.py (hunk 13 above)
Comment (vinniefalco) [inline bin/manifest.py]: It matches the order in the rippled C++ sources

File: bin/manifest.py (hunk 27 above)
Comment (vinniefalco) [inline bin/manifest.py]: You need at least `sk`

**Comment by vinniefalco** [comment]
The Python needs work

**Comment by rec** [comment]
would you like to pull it out as a separate change?

I could take it and massage it...?

**Comment by vinniefalco** [comment]
@rec Yes I think that would be best. Just start a new branch off b22, manually copy the manifest.py program and then use your best judgement for the 3rd party libraries.

**Comment by rec** [comment]
OK.  On it almost immediately.

File: src/ripple/basics/strHex.h [hunk 34]
```diff
@@ -25,6 +25,7 @@
 #ifndef RIPPLE_BASICS_STRHEX_H_INCLUDED
 #define RIPPLE_BASICS_STRHEX_H_INCLUDED
 
+#include <cassert>
```
Comment (ximinez) [inline src/ripple/basics/strHex.h]: I always get suspicious when I see `#include`s added with no other code changes to that file.

File: src/ripple/overlay/impl/Manifest.cpp [hunk 35]
```diff
@@ -0,0 +1,247 @@
...
+unpackManifest(std::string s)
+{
+    STObject st(sfGeneric);
+    SerialIter sit(s.data(), s.size());
+    st.set(sit);
+    auto const mseq = get(st, sfSequence);
+    auto mpk  = get<AnyPublicKey>(st, sfPublicKey);
+    auto mspk = get<AnyPublicKey>(st, sfSigningPubKey);
+    if (! mseq || ! mpk || ! mspk)
+        return boost::optional<Manifest>();
```
Comment (ximinez) [inline src/ripple/overlay/impl/Manifest.cpp]: Here and anywhere else you return a default created `boost::optional<T>()`, you can use the simpler shorthand of `boost::none`.

File: src/ripple/basics/strHex.h (hunk 34 above)
Comment (nbougalis) [inline src/ripple/basics/strHex.h]: I suspect this is needed when compiling non-unity.

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 36]
```diff
@@ -566,10 +605,77 @@ OverlayImpl::onPeerDeactivate (Peer::id_t id,
...
+        if (! pk ||
+            ! seq ||
+            ! get(st, sfSignature) ||
+            ! get(st, sfSigningPubKey))           
+        {
+            if (journal.info) journal.info
+                << "Malformed manifest #" << i + 1;
+            continue;
+        }
+        // `continue` if pk is untrusted
```
Comment (ximinez) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: This comment doesn't really add anything. And it's a little misleading, because `maybe_accept` can fail for other reasons.

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 37]
```diff
@@ -566,10 +605,77 @@ OverlayImpl::onPeerDeactivate (Peer::id_t id,
...
+        if (! manifestCache_.maybe_accept (*pk, *seq, s, st, journal))
+        {
+            if (journal.warning) journal.warning
+                << "Ignored manifest seq #" << *seq << " for " << *pk;
+            continue;
+        }
+        if (journal.warning) journal.warning
```
Comment (ximinez) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: Per recent discussion, I don't think the ignore/accept/forward messages should be logged at warning level. Maybe one step down.

File: src/ripple/protocol/AnyPublicKey.h [hunk 38]
```diff
@@ -102,9 +102,16 @@ class AnyPublicKey
...
-    AnyPublicKey (AnyPublicKey const&) = delete;
     AnyPublicKey& operator= (AnyPublicKey const&) = delete;
 
+    AnyPublicKey (AnyPublicKey const& other)
+        : buffer_type(other.buffer_type::member.data(),
+            other.buffer_type::member.size())
+        , AnyPublicKeySlice (buffer_type::member.data(),
+            buffer_type::member.size())
+    {
+    }
```
Comment (ximinez) [inline src/ripple/protocol/AnyPublicKey.h]: This may be a duplicate comment, but if you provide the copy constructor, you should also provide the copy assignment, or include a comment explaining why you aren't.

File: bin/manifest.py (hunk 18 above)
Comment (josh-ripple) [inline bin/manifest.py]: I grant you `'foo'` vs. `"foo"`, but `''` can be confused for `' '`, a problem not shared with `""` and `" "`

File: src/ripple/overlay/tests/manifest_test.cpp [hunk 39]
```diff
@@ -0,0 +1,107 @@
...
+        expect(! cache.would_accept(pk_a, 0));
+        expect(! cache.maybe_insert(pk_a, 0, s_a0.first, journal));
+
+        expect(cache.would_accept(pk_b, 0));
+        expect(cache.maybe_insert(pk_b, 0, s_b0.first, journal));
+
+        expect(! cache.maybe_accept(pk_b, 0, s_b0.first, s_b0.second, journal));
+        expect(cache.maybe_accept(pk_b, 1, s_b1.first, s_b1.second, journal));
+
+        expect(! cache.maybe_accept(pk_b, 2, s_b0.first, s_a0.second, journal), "wrong sig not accepted");
```
Comment (ximinez) [inline src/ripple/overlay/tests/manifest_test.cpp]: Nice tests!

File: src/ripple/protocol/AnyPublicKey.h (hunk 38 above)
Comment (vinniefalco) [inline src/ripple/protocol/AnyPublicKey.h]: The existence of a copy constructor implies we are making needless copies somewhere. If this copy constructor is commented out, which line then breaks?

File: src/ripple/basics/strHex.h (hunk 34 above)
Comment (ximinez) [inline src/ripple/basics/strHex.h]: Bingo.

Comment (josh-ripple) [inline src/ripple/basics/strHex.h]: True.  That header is required but is already included earlier in the translation unit… for now.   A subsequent commit changes that, and the missing include is now necessary.  Since it belonged there in the first place I treated it as a fix and committed it separately.

File: src/ripple/overlay/impl/Manifest.cpp (hunk 35 above)
Comment (josh-ripple) [inline src/ripple/overlay/impl/Manifest.cpp]: Ooh, good idea.  Done.

File: src/ripple/overlay/impl/OverlayImpl.cpp (hunk 36 above)
Comment (josh-ripple) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: Okay, it's gone.

**Comment by scottschurr** [comment]
Is there more work to do here?  Or is it time for a squash and a rebase?

**Comment by ximinez** [comment]
I left comments related to `AnyPublicKey` copy constructor and logging levels which I don't see have been addressed yet.

File: src/ripple/overlay/impl/OverlayImpl.cpp (hunk 37 above)
Comment (josh-ripple) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: I've audited the diagnostics and downgraded most of them to `info` or `debug`, removing the redundant ones.  I documented the remaining ones except for the two warning about missing config sections, which are obvious as is.

File: src/ripple/protocol/AnyPublicKey.h (hunk 38 above)
Comment (josh-ripple) [inline src/ripple/protocol/AnyPublicKey.h]: (a) Copy construction is needed, but copy assignment isn't.  (b) I tried to implement copy assignment and failed — I'm not sure `Buffer` allows for it.  I'd prefer to leave something more substantive than my uncertainty in a code comment… ideas?

Comment (vinniefalco) [inline src/ripple/protocol/AnyPublicKey.h]: If this copy constructor is commented out, which line then breaks?

Comment (josh-ripple) [inline src/ripple/protocol/AnyPublicKey.h]: `unl.insertEphemeralKey (m->signingKey, iter->second.comment);`

It needs to copy the signing key.  Unless you want to risk storing an `AnyPublicKeySlice`…

Another way to save a memory allocation would be to have `AnyPublicKey` objects share the buffer (since they're inherently immutable values) but I don't think it's worth it.  This is not a performance or maintenance issue and we're grinding our own dust.  This inquiry has already resulted in a new error in the code (since fixed).

Comment (vinniefalco) [inline src/ripple/protocol/AnyPublicKey.h]: We definitely don't want to store a slice in `UniqueNodeList`. I think this is sufficient evidence that we really do need a copy constructor (because we have to put it in two places).

**Comment by josh-ripple** [comment]
This pull request is superseded by <https://github.com/ripple/rippled/pull/990>.

**Comment by nbougalis** [comment]
I am closing this PR since it's been superseded. Existing comments will remain here for posterity. Use the new PR for any comments going forward.

File: src/ripple/overlay/impl/PeerImp.cpp [hunk 40]
```diff
@@ -714,6 +730,15 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMHello> const& m)
...
 
 void
+PeerImp::onMessage (std::shared_ptr<protocol::TMManifests> const& m)
+{
+    // VFALCO What's the right job type?
+    getApp().getJobQueue().addJob (jtVALIDATION_ut,
+        "receiveManifests", std::bind(&OverlayImpl::onManifests,
+            &overlay_, std::placeholders::_1, m, shared_from_this()));
+}
+
```
Comment (JoelKatz) [inline src/ripple/overlay/impl/PeerImp.cpp]: This is significantly inferior to the way we handle other signed objects like transactions, validations, and proposals. I'm concerned that we'll wind up with all our threads blocked in these jobs, waiting for the manifest lock which is held while we check the signature.

At an absolute minimum, dispatch these as their own job type with a concurrency limit of one. But I'd much rather see the same mechanism we use for other such objects used for those. For example, just by using the HashRouter, we can avoid dispatching every copy of a manifest we receive and we can avoid broadcasting the manifest to servers we receive it from.
